pkg/controller/volume/persistentvolume/scheduler_binder_test.go

```diff
@@ -1302,63 +1431,110 @@ func TestCheckBindings(t *testing.T) {
-		nodes           []*v1.Node
+		bindingsNil bool // Pass in nil bindings slice
+
+		nodes []*v1.Node
+
+		// before assume
+		initPVs  []*v1.PersistentVolume
+		initPVCs []*v1.PersistentVolumeClaim
+
+		// assume PV & PVC with these binding results
+		binding          *bindingInfo
+		claimToProvision *v1.PersistentVolumeClaim
+
+		// API updates after assume before bind
+		apiPV  *v1.PersistentVolume
+		apiPVC *v1.PersistentVolumeClaim
 
 		// This function runs with a delay of 5 seconds
-		delayFunc func(*testing.T, *testEnv, *v1.Pod, *bindingInfo, []*v1.PersistentVolumeClaim)
+		delayFunc func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pvs []*v1.PersistentVolume, pvcs []*v1.PersistentVolumeClaim)
 
 		// Expected return values
 		shouldFail bool
-	}{
+	}
+
+	scenarios := map[string]scenarioType{
 		"nothing-to-bind-nil": {
 			bindingsNil: true,
 			shouldFail:  true,
 		},
 		"nothing-to-bind-empty": {},
 		"already-bound": {
-			binding:    binding1aBound,
-			cachedPVs:  []*v1.PersistentVolume{pvNode1aBound},
-			cachedPVCs: []*v1.PersistentVolumeClaim{boundPVCNode1a},
+			binding:  makeBinding(unboundPVC, pvNode1aBound),
+			initPVs:  []*v1.PersistentVolume{pvNode1aBound},
+			initPVCs: []*v1.PersistentVolumeClaim{boundPVCNode1a},
 		},
 		"binding-succeeds-after-time": {
-			binding:    binding1aBound,
-			cachedPVs:  []*v1.PersistentVolume{pvNode1a},
-			cachedPVCs: []*v1.PersistentVolumeClaim{unboundPVC},
-			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, binding *bindingInfo, pvcs []*v1.PersistentVolumeClaim) {
+			initPVs:    []*v1.PersistentVolume{pvNode1aBound},
+			initPVCs:   []*v1.PersistentVolumeClaim{unboundPVC},
+			binding:    makeBinding(unboundPVC, pvNode1aBound),
+			shouldFail: false, // Will succeed after PVC is fully bound to this PV by pv controller.
+			delayFunc: func(t *testing.T, testEnv *testEnv, pod *v1.Pod, pvs []*v1.PersistentVolume, pvcs []*v1.PersistentVolumeClaim) {
+				pvc := pvcs[0]
+				pv := pvs[0]
 				// Update PVC to be fully bound to PV
-				newPVC := binding.pvc.DeepCopy()
-				newPVC.ResourceVersion = "100"
-				newPVC.Spec.VolumeName = binding.pv.Name
+				newPVC := pvc.DeepCopy()
+				newPVC.Spec.VolumeName = pv.Name
 				metav1.SetMetaDataAnnotation(&newPVC.ObjectMeta, annBindCompleted, "yes")
-
-				// Update pvc cache, fake client doesn't invoke informers
-				internalBinder, ok := testEnv.binder.(*volumeBinder)
-				if !ok {
-					t.Fatalf("Failed to convert to internal binder")
+				testEnv.client.CoreV1().PersistentVolumeClaims(newPVC.Namespace).Update(newPVC)
+			},
+		},
+		"bound-by-pv-controller-before-bind": {
+			initPVs:    []*v1.PersistentVolume{pvNode1a},
+			initPVCs:   []*v1.PersistentVolumeClaim{unboundPVC},
+			binding:    makeBinding(unboundPVC, pvNode1aBound),
+			apiPV:      pvNode1aBound,
+			apiPVC:     boundPVCNode1a,
+			shouldFail: true, // bindAPIUpdate will fail because API conflict
+		},
+		"binding-static-pv-succeeds-after-time": {
```

Comment:
What's the difference between this and "binding-succeeds-after-time"? That the PV object is not bound?

Reply:
they are same... my mistake, i'm going to remove one of them.